File: dotnet/M3Mobile.M3Sdk.Xamarin/KeyTool/KeyToolVersionPolicy.cs
using System;
using M3Sdk.Xamarin.Internal;

namespace M3Sdk.Xamarin.KeyTool
{
    internal static class KeyToolVersionPolicy
    {
        internal const string FnMinimumVersion = "1.2.6";
        internal const string NavigationMinimumVersion = "1.4.1";
        internal const string ScanWakeUpMinimumVersion = "1.3.8";

        internal static string KeyFunctionMinimumVersion(DeviceModel model)
        {
            if (model == DeviceModel.SM24)
                return "1.3.8";
            if (model == DeviceModel.SM25)
                return "1.3.16";
            return "1.2.6";
        }

        internal static bool VersionSatisfied(string currentVersion, string requiredVersion)
        {
            var current = VersionParts(currentVersion);
            var required = VersionParts(requiredVersion);

            for (var i = 0; i < current.Length; i++)
            {
                if (current[i] != required[i])
                    return current[i] > required[i];
            }

            return true;
        }

        private static int[] VersionParts(string version)
        {
            var parts = new[] { 0, 0, 0 };
            if (string.IsNullOrEmpty(version))
                return parts;

            var tokens = version.Split('.');
            for (var i = 0; i < parts.Length && i < tokens.Length; i++)
                parts[i] = NumericPrefix(tokens[i]);

            return parts;
        }

        private static int NumericPrefix(string token)
        {
            var length = 0;
            while (length < token.Length && char.IsDigit(token[length]))
                length++;

            int value;
            return length > 0 && int.TryParse(token.Substring(0, length), out value)
                ? value
                : 0;
        }
    }
}

File: feature/keytool/src/main/java/net/m3mobile/feature/keytool/internal/KeyToolVersionPolicy.kt
package net.m3mobile.feature.keytool.internal

import net.m3mobile.core.device.DeviceModel

internal object KeyToolVersionPolicy {

    const val FN_MINIMUM_VERSION = "1.2.6"
    const val NAVIGATION_MINIMUM_VERSION = "1.4.1"
    const val SCAN_WAKE_UP_MINIMUM_VERSION = "1.3.8"

    fun keyFunctionMinimumVersion(model: DeviceModel): String = when (model) {
        DeviceModel.SM24 -> "1.3.8"
        DeviceModel.SM25 -> "1.3.16"
        else -> "1.2.6"
    }

    fun versionSatisfied(currentVersion: String, requiredVersion: String): Boolean {
        val current = versionParts(currentVersion)
        val required = versionParts(requiredVersion)

        return current.zip(required)
            .firstOrNull { (currentPart, requiredPart) -> currentPart != requiredPart }
            ?.let { (currentPart, requiredPart) -> currentPart > requiredPart }
            ?: true
    }

    private fun versionParts(version: String): List<Int> {
        val tokens = version.split('.')
        return List(VERSION_PART_COUNT) { index ->
            tokens.getOrNull(index)
                ?.takeWhile(Char::isDigit)
                ?.toIntOrNull()
                ?: 0
        }
    }

    private const val VERSION_PART_COUNT = 3
}
